src/Razor/src/Microsoft.CodeAnalysis.Remote.Razor/InlayHints/RemoteInlayHintService.cs

```diff
@@ -0,0 +1,115 @@
+       => RunServiceAsync(
+            solutionInfo,
+            razorDocumentId,
+            context => GetInlayHintsAsync(context, inlayHintParams, displayAllOverride, cancellationToken),
+            cancellationToken);
+
+    private async ValueTask<InlayHint[]?> GetInlayHintsAsync(RemoteDocumentContext context, InlayHintParams inlayHintParams, bool displayAllOverride, CancellationToken cancellationToken)
+    {
+        var codeDocument = await context.GetCodeDocumentAsync(cancellationToken).ConfigureAwait(false);
+        var csharpDocument = codeDocument.GetCSharpDocument();
+
+        var span = inlayHintParams.Range.ToLinePositionSpan();
+
+        // We are given a range by the client, but our mapping only succeeds if the start and end of the range can both be mapped
+        // to C#. Since that doesn't logically match what we want from inlay hints, we instead get the minimum range of mappable
+        // C# to get hints for. We'll filter that later, to remove the sections that can't be mapped back.
+        if (!_documentMappingService.TryMapToGeneratedDocumentRange(csharpDocument, span, out var projectedLinePositionSpan) &&
+            !codeDocument.TryGetMinimalCSharpRange(span, out projectedLinePositionSpan))
+        {
+            // There's no C# in the range.
+            return null;
+        }
+
+        var generatedDocument = await context.GetGeneratedDocumentAsync(_filePathService, cancellationToken).ConfigureAwait(false);
+
+        var textDocument = inlayHintParams.TextDocument.WithUri(generatedDocument.CreateUri());
+        var range = projectedLinePositionSpan.ToRange();
+
+        var hints = await InlayHints.GetInlayHintsAsync(generatedDocument, textDocument, range, displayAllOverride, cancellationToken).ConfigureAwait(false);
+
+        if (hints is null)
+        {
+            return null;
+        }
+
+        using var inlayHintsBuilder = new PooledArrayBuilder<InlayHint>();
+        var razorSourceText = codeDocument.Source.Text;
+        var csharpSourceText = codeDocument.GetCSharpSourceText();
+        var syntaxTree = codeDocument.GetSyntaxTree();
+        foreach (var hint in hints)
+        {
+            if (csharpSourceText.TryGetAbsoluteIndex(hint.Position.ToLinePosition(), out var absoluteIndex) &&
+                _documentMappingService.TryMapToHostDocumentPosition(csharpDocument, absoluteIndex, out var hostDocumentPosition, out var hostDocumentIndex))
+            {
+                // We know this C# maps to Razor, but does it map to Razor that we like?
+                var node = syntaxTree.Root.FindInnermostNode(hostDocumentIndex);
+                if (node?.FirstAncestorOrSelf<MarkupTagHelperAttributeValueSyntax>() is not null)
+                {
+                    continue;
+                }
```

Comment:
Annoyingly, this bit of code is pretty much the only logic that could be shared with the existing inlay hints code, because everything else operates on VS LSP types, and here we're using Roslyn LSP types. The code duplication isn't great, but I'm somewhat okay with it because we have good tests for both.

Perhaps we can re-review if/when we can move the language server to Roslyn LSP types, which could be as early as preview 2.